include/wait_strategy.hpp

```diff
@@ -0,0 +1,379 @@
+ * It is best used when threads can be bound to specific CPU cores.
+ */
+struct BusySpinWaitStrategy {
+    std::int64_t waitFor(const std::int64_t sequence, const Sequence & /*cursor*/, const std::vector<std::shared_ptr<Sequence>> &dependentSequences) const {
+        std::int64_t availableSequence;
+        while ((availableSequence = detail::getMinimumSequence(dependentSequences)) < sequence) {
+            // optional: barrier check alert
+        }
+        return availableSequence;
+    }
+};
+static_assert(WaitStrategy<BusySpinWaitStrategy>);
+static_assert(!hasSignalAllWhenBlocking<BusySpinWaitStrategy>);
+
+/**
+ * Sleeping strategy that initially spins, then uses a std::this_thread::yield(), and eventually sleep. T
+ * his strategy is a good compromise between performance and CPU resource.
+ * Latency spikes can occur after quiet periods.
+ */
+class SleepingWaitStrategy {
+    static const std::int32_t _defaultRetries = 200;
+    std::int32_t              _retries        = 0;
+
+public:
+    explicit SleepingWaitStrategy(std::int32_t retries = _defaultRetries)
+        : _retries(retries) {
+    }
+
+    std::int64_t waitFor(const std::int64_t sequence, const Sequence & /*cursor*/, const std::vector<std::shared_ptr<Sequence>> &dependentSequences) const {
+        auto       counter    = _retries;
+        const auto waitMethod = [&]() {
+            // optional: barrier check alert
+
+            if (counter > 100) {
+                --counter;
+            } else if (counter > 0) {
+                --counter;
+                std::this_thread::yield();
+            } else {
+                std::this_thread::sleep_for(std::chrono::milliseconds(0));
+            }
+
+            return counter;
+        };
+
+        std::int64_t availableSequence;
+        while ((availableSequence = detail::getMinimumSequence(dependentSequences)) < sequence) {
+            counter = waitMethod();
+        }
+
+        return availableSequence;
+    }
+};
+static_assert(WaitStrategy<SleepingWaitStrategy>);
+static_assert(!hasSignalAllWhenBlocking<SleepingWaitStrategy>);
+
+struct TimeoutException : public std::runtime_error {
+    TimeoutException() : std::runtime_error("TimeoutException") {}
+};
+
+class TimeoutBlockingWaitStrategy {
+    using Clock = std::conditional_t<std::chrono::high_resolution_clock::is_steady, std::chrono::high_resolution_clock, std::chrono::steady_clock>;
+    Clock::duration             _timeout;
+    std::recursive_mutex        _gate;
+    std::condition_variable_any _conditionVariable;
+
+public:
+    explicit TimeoutBlockingWaitStrategy(Clock::duration timeout)
+        : _timeout(timeout) {}
+
+    std::int64_t waitFor(const std::int64_t sequence, const Sequence &cursor, const std::vector<std::shared_ptr<Sequence>> &dependentSequences) {
+        auto timeSpan = std::chrono::microseconds(std::chrono::duration_cast<std::chrono::microseconds>(_timeout).count());
```

Comment:
I don't get this - cast to microseconds, then .count, then construct microseconds?

Reply:
Just an elaborate scheme to confuse reviewers... ;-)
removed it.